System degrades if an empty entry is created

If a user tries to get data from not existing entry, the storage created it. After that, the quota stops working, and it creates many logs and 
stop working  at all. 

```
 [ERROR] -- storage.cc:199 Didn't mange to keep quota: [500] Tries to remove a block in empty entry 
```

The solution:

1. Fix quota not to clean blocks from already empty entries.
2. Not create an entry for GET request

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,5 +1,9 @@
 ## Release 0.6.0 (in progress)
 
+## Release 0.5.1 (2022-05-24)
+
+**Bugs**:
+* GET /b/:bucket/:entry doesn't create an empty entry, [PR-95](https://github.com/reduct-storage/reduct-storage/pull/95)
 ## Release 0.5.0 (2022-05-15)
 
 **Features**:

--- a/api_tests/entry_api_test.py
+++ b/api_tests/entry_api_test.py
@@ -57,7 +57,7 @@ def test_read_no_data(base_url, headers, bucket):
 
 def test_read_bad_ts(base_url, headers, bucket):
     """Should return 400 if ts is bad"""
-    requests.post(f"{base_url}/b/{bucket}/entry?ts=100", data=b"somedata")
+    requests.post(f"{base_url}/b/{bucket}/entry?ts=100", data=b"somedata", headers=headers)
     resp = requests.get(f'{base_url}/b/{bucket}/entry?ts=XXXX', headers=headers)
     assert resp.status_code == 422
     assert 'XXX' in get_detail(resp)

--- a/api_tests/entry_api_test.py
+++ b/api_tests/entry_api_test.py
@@ -57,9 +57,17 @@ def test_read_no_data(base_url, headers, bucket):
 
 def test_read_bad_ts(base_url, headers, bucket):
     """Should return 400 if ts is bad"""
+    requests.post(f"{base_url}/b/{bucket}/entry?ts=100", data=b"somedata")
     resp = requests.get(f'{base_url}/b/{bucket}/entry?ts=XXXX', headers=headers)
     assert resp.status_code == 422
     assert 'XXX' in get_detail(resp)
+
+
+def test_read_bad_no_entry(base_url, headers, bucket):
+    """Should return 400 if ts is bad"""
+    resp = requests.get(f'{base_url}/b/{bucket}/entry', headers=headers)
+    assert resp.status_code == 404
+    assert 'entry' in get_detail(resp)
 
 
 def test_write_no_bucket(base_url, headers):

--- a/docs/SUMMARY.md
+++ b/docs/SUMMARY.md
@@ -11,10 +11,7 @@
 ## SDKs
 
 * [Python Client SDK](https://reduct-py.readthedocs.io)
-<<<<<<< HEAD
-=======
 * [JavaScript Client SDK](https://reduct-js.readthedocs.io/en/latest/)
->>>>>>> main
 * [C++ Client SDK](https://reduct-cpp.readthedocs.io/en/latest/)
 
 ## VERSIONS

--- a/docs/http-api/entry-api.md
+++ b/docs/http-api/entry-api.md
@@ -21,7 +21,7 @@ Name of entry
 A UNIX timestamp in microseconds
 {% endswagger-parameter %}
 
-{% swagger-parameter in="header" required="true" name="size" %}
+{% swagger-parameter in="header" required="true" name="Content-Length" %}
 Content-length is required to start asynchronous write operation
 {% endswagger-parameter %}
 

--- a/src/reduct/api/common.h
+++ b/src/reduct/api/common.h
@@ -131,6 +131,8 @@ class BasicApiHandler {
   void SendError(core::Error err) const noexcept {
     if (err.code >= 500) {
       LOG_ERROR("{} {}: {}", method_, url_, err.ToString());
+    } else {
+      LOG_DEBUG("{} {}: {}", method_, url_, err.ToString());
     }
     http_resp_->writeStatus(std::to_string(err.code));
     http_resp_->end(fmt::format(R"({{"detail":"{}"}})", err.message));

--- a/src/reduct/storage/bucket.cc
+++ b/src/reduct/storage/bucket.cc
@@ -155,6 +155,10 @@ class Bucket : public IBucket {
     return std::vector(std::ranges::begin(rr), std::ranges::end(rr));
   }
 
+  bool HasEntry(const std::string& name) const override {
+    return entry_map_.contains(name);
+  }
+
   [[nodiscard]] BucketInfo GetInfo() const override {
     size_t record_count = 0;
     size_t size = 0;

--- a/src/reduct/storage/bucket.cc
+++ b/src/reduct/storage/bucket.cc
@@ -112,6 +112,7 @@ class Bucket : public IBucket {
   }
 
   [[nodiscard]] Error KeepQuota() override {
+    auto err = Error::kOk;
     switch (settings_.quota_type()) {
       case BucketSettings::NONE:
         break;
@@ -127,22 +128,24 @@ class Bucket : public IBucket {
           std::vector<EntryInfo> candidates(std::ranges::begin(rr), std::ranges::end(rr));
           std::ranges::sort(candidates, [](auto& a, auto& b) { return a.oldest_record() < b.oldest_record(); });
 
-          const auto& entry = candidates[0];
-          LOG_DEBUG("Remove the oldest block in entry '{}'", entry.name());
-          auto err = entry_map_.at(entry.name())->RemoveOldestBlock();
-          if (err) {
-            return err;
-          }
+          for (int i = 0; i < candidates.size(); ++i) {
+            const auto& entry = candidates[i];
+            LOG_DEBUG("Remove the oldest block in entry '{}'", entry.name());
+            auto entry_ptr = entry_map_.at(entry.name());
+            err = entry_ptr->RemoveOldestBlock();
+            if (!err) {
+              if (entry_ptr->GetInfo().block_count() == 0) {
+                entry_map_.erase(entry.name());
+                fs::remove(full_path_ / entry.name());
+              }
 
-          if (entry_map_.at(entry.name())->GetInfo().block_count() == 0) {
-            entry_map_.erase(entry.name());
-            fs::remove(full_path_ / entry.name());
+              bucket_size = GetInfo().size();
+              break;
+            }
           }
-
-          bucket_size = GetInfo().size();
         }
     }
-    return Error::kOk;
+    return err;
   }
 
   Error SetSettings(BucketSettings settings) override {
@@ -155,9 +158,7 @@ class Bucket : public IBucket {
     return std::vector(std::ranges::begin(rr), std::ranges::end(rr));
   }
 
-  bool HasEntry(const std::string& name) const override {
-    return entry_map_.contains(name);
-  }
+  bool HasEntry(const std::string& name) const override { return entry_map_.contains(name); }
 
   [[nodiscard]] BucketInfo GetInfo() const override {
     size_t record_count = 0;

--- a/src/reduct/storage/bucket.h
+++ b/src/reduct/storage/bucket.h
@@ -69,11 +69,15 @@ class IBucket {
    * @return
    */
   virtual std::vector<proto::api::EntryInfo> GetEntryList() const = 0;
+
+
+  virtual bool HasEntry(const std::string& name) const = 0;
   /**
    * @brief Builds a new bucket
    * @param options
    * @return
    */
+
   static std::unique_ptr<IBucket> Build(std::filesystem::path full_path, proto::api::BucketSettings options = {});
 
   /**

--- a/src/reduct/storage/storage.cc
+++ b/src/reduct/storage/storage.cc
@@ -212,7 +212,13 @@ class Storage : public IStorage {
         return Callback::Result{{}, err};
       }
 
-      auto [entry, ref_error] = bucket_it->second->GetOrCreateEntry(std::string(req.entry_name));
+      auto& bucket = bucket_it->second;
+      if (!bucket->HasEntry(req.entry_name.data())) {
+        return Callback::Result{{},
+                                {.code = 404, .message = fmt::format("Entry '{}' could not be found", req.entry_name)}};
+      }
+
+      auto [entry, ref_error] = bucket->GetOrCreateEntry(req.entry_name.data());
       if (ref_error) {
         return Callback::Result{{}, ref_error};
       }

--- a/unit_tests/reduct/storage/bucket_test.cc
+++ b/unit_tests/reduct/storage/bucket_test.cc
@@ -138,6 +138,15 @@ TEST_CASE("storage::Bucket should keep quota", "[bucket]") {
 
     REQUIRE(bucket->GetEntryList().size() == 1);
   }
+
+  SECTION("should work with empty entry-2") {
+    REQUIRE(entry1->BeginWrite(ts + seconds(1), blob.size()).result->Write(blob) == Error::kOk);
+    REQUIRE(entry1->BeginWrite(ts + seconds(2), blob.size()).result->Write(blob) == Error::kOk);
+    REQUIRE(entry1->BeginWrite(ts + seconds(3), blob.size()).result->Write(blob) == Error::kOk);
+
+    REQUIRE(bucket->KeepQuota() == Error::kOk);
+    REQUIRE(entry1->GetInfo().record_count() == 2);
+  }
 }
 
 TEST_CASE("storage::Bucket should change quota settings and save it", "[bucket]") {

--- a/unit_tests/reduct/storage/storage_entry_api_test.cc
+++ b/unit_tests/reduct/storage/storage_entry_api_test.cc
@@ -23,6 +23,7 @@ using reduct::storage::IStorage;
 
 using reduct::MakeDefaultBucketSettings;
 using reduct::OnCreateBucket;
+using reduct::OnGetBucket;
 using reduct::OnListEntry;
 using reduct::OnReadEntry;
 using reduct::OnWriteEntry;
@@ -101,6 +102,16 @@ TEST_CASE("storage::Storage should write and read data", "[storage][entry]") {
     error = OnReadEntry(storage.get(), {.bucket_name = "bucket", .entry_name = "entry", .timestamp = "XXXX"}).Get();
     REQUIRE(error ==
             Error{.code = 422, .message = "Failed to parse 'ts' parameter: XXXX should unix times in microseconds"});
+  }
+
+  SECTION("error if no entry to read") {
+    Error error =
+        OnReadEntry(storage.get(), {.bucket_name = "bucket", .entry_name = "NOTEXIT", .timestamp = "1000"}).Get();
+    REQUIRE(error == Error{.code = 404, .message = "Entry 'NOTEXIT' could not be found"});
+
+    auto bucket_info = OnGetBucket(storage.get(), {.bucket_name="bucket"}).Get();
+    REQUIRE(bucket_info == Error::kOk);
+    REQUIRE(bucket_info.result.entries_size() == 1);  // we don't create a new one
   }
 
   SECTION("error if the record not found") {

--- a/unit_tests/reduct/storage/storage_entry_api_test.cc
+++ b/unit_tests/reduct/storage/storage_entry_api_test.cc
@@ -109,7 +109,7 @@ TEST_CASE("storage::Storage should write and read data", "[storage][entry]") {
         OnReadEntry(storage.get(), {.bucket_name = "bucket", .entry_name = "NOTEXIT", .timestamp = "1000"}).Get();
     REQUIRE(error == Error{.code = 404, .message = "Entry 'NOTEXIT' could not be found"});
 
-    auto bucket_info = OnGetBucket(storage.get(), {.bucket_name="bucket"}).Get();
+    auto bucket_info = OnGetBucket(storage.get(), {.bucket_name = "bucket"}).Get();
     REQUIRE(bucket_info == Error::kOk);
     REQUIRE(bucket_info.result.entries_size() == 1);  // we don't create a new one
   }
@@ -176,7 +176,7 @@ TEST_CASE("storage::Storage should list records by timestamps", "[storage][entry
   }
 
   SECTION("error, if bucket not found") {
-    REQUIRE(OnListEntry(storage.get(), {.bucket_name = "UNKNOWN_BUCKET",
+     REQUIRE(OnListEntry(storage.get(), {.bucket_name = "UNKNOWN_BUCKET",
                                         .entry_name = "entry",
                                         .start_timestamp = "1000",
                                         .stop_timestamp = "1200"})

